platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AppendOnceList.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import com.google.common.base.Preconditions;
+
+import javax.annotation.CheckReturnValue;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.stream.Collector;
+import java.util.stream.Stream;
+
+/**
+ * A special append-only list used in collection properties implementations.
+ * Pretends to be an immutable data structure but isn't actually thread-safe.
+ * <p>
+ * Appending an element to the list can only be done once, because the internal buffer is shared by this list and the new one.
```

Comment:
```suggestion
 * Appending an element to each `AppendOnceList` (via {@link #plus}) can only be done once, because the internal buffer is shared by this list and the new one. Chained calls to `plus` are ok, e.g., `AppendOnceList<String>().plus("foo").plus("bar")`.
```